File: app-runtime-management/agents/macos/Sources/MacOSAppRuntimeCore/UploadOutbox.swift
public struct UploadOutboxEntry: Codable, Equatable, Sendable {
    public let segmentID: String
    public let attemptCount: Int
    public let nextAttemptAtMs: Int64?
    public let lastErrorCode: String?

    public init(
        segmentID: String,
        attemptCount: Int,
        nextAttemptAtMs: Int64? = nil,
        lastErrorCode: String? = nil
    ) {
        self.segmentID = segmentID
        self.attemptCount = attemptCount
        self.nextAttemptAtMs = nextAttemptAtMs
        self.lastErrorCode = lastErrorCode
    }
}

public struct UploadRejection: Codable, Equatable, Sendable {
    public let id: String
    public let code: String

    public init(id: String, code: String) {
        self.id = id
        self.code = code
    }
}

public struct UploadAcceptance: Codable, Equatable, Sendable {
    public let acceptedIds: [String]
    public let rejected: [UploadRejection]

    public init(acceptedIds: [String], rejected: [UploadRejection]) {
        self.acceptedIds = acceptedIds
        self.rejected = rejected
    }
}

public protocol UploadOutbox: Sendable {
    func pending(limit: Int, nowMs: Int64) async throws -> [UploadOutboxEntry]
    func markAccepted(segmentIDs: Set<String>) async throws
    func recordFailure(
        segmentIDs: Set<String>,
        errorCode: String,
        retryAtMs: Int64
    ) async throws
}

File: app-runtime-management/agents/windows/src/TimeOnChrome.AppRuntime.Core/UploadOutbox.cs
namespace TimeOnChrome.AppRuntime.Core;

public sealed record UploadOutboxEntry(
    string SegmentID,
    int AttemptCount,
    long? NextAttemptAtMs = null,
    string? LastErrorCode = null);

public sealed record UploadRejection(string Id, string Code);

public sealed record UploadAcceptance(
    IReadOnlyList<string> AcceptedIds,
    IReadOnlyList<UploadRejection> Rejected);

public interface IUploadOutbox
{
    Task<IReadOnlyList<UploadOutboxEntry>> PendingAsync(
        int limit,
        long nowMs,
        CancellationToken cancellationToken = default);

    Task MarkAcceptedAsync(
        IReadOnlySet<string> segmentIDs,
        CancellationToken cancellationToken = default);

    Task RecordFailureAsync(
        IReadOnlySet<string> segmentIDs,
        string errorCode,
        long retryAtMs,
        CancellationToken cancellationToken = default);
}
